### PR TITLE
Only add big objarray to remset once

### LINKED ISSUE #49205: The new marking loop has a regression when marking arrays of pointers
When looking into https://github.com/JuliaLang/julia/issues/49120, it seems the regression seen there was hiding a part of the new mark loop where it's slower when marking a very large array of pointers. 

```julia
using Random: seed!
seed!(1)

abstract type Cell end

struct CellA<:Cell
    a::Int
end

struct CellB<:Cell
    b::String
end

function fillcells!(mc::Array{Cell})
    for ind in eachindex(mc)
        mc[ind] = ifelse(rand() > 0.5, CellA(ind), CellB(string(ind)))
    end
    return mc
end

mcells = Array{Cell}(undef, 5000, 5000 )
t1 = @elapsed fillcells!(mcells)
t2 = @elapsed fillcells!(mcells)

println("filling: $t1 s\nfilling again: $t2 s")

@time GC.gc()
@time GC.gc()
```
One of the GCs after will do a full mark of the array and take a long time to do it. This probably wasn't seen before because arrays of many pointers had really bad GC behaviour overall, but with https://github.com/JuliaLang/julia/pull/49185 it's quite clear.

### PR DESCRIPTION
**Pull request by gbaraldi**
Fixes https://github.com/JuliaLang/julia/issues/49205

### CODE REVIEW

**Review by vtjnash — APPROVED** [review]

File: src/gc.c [hunk 1]
```diff
@@ -2082,15 +2082,37 @@ STATIC_INLINE void gc_mark_objarray(jl_ptls_t ptls, jl_value_t *obj_parent, jl_v
...
+                    break;
+                gc_heap_snapshot_record_array_edge(obj_parent, &new_obj);
+            }
+        }
+    }
+    size_t too_big = (obj_end - obj_begin) / MAX_REFS_AT_ONCE > step; // use this order of operations to avoid idiv
+    jl_value_t **scan_end = obj_end;
+    if (too_big) {
+        scan_end = obj_begin + step * MAX_REFS_AT_ONCE;
+        if ((nptr & 0x3) != 0x2) {
```
Comment (d-netto) [inline src/gc.c:2107]: Why are we creating a chunk conditioned on `(nptr & 0x3) != 0x2`?

**Comment by d-netto** [comment]
What's the performance impact of this PR on https://github.com/JuliaLang/julia/issues/49205?

**Comment by gbaraldi** [comment]
Makes it faster than on 1.8

**Review by d-netto — CHANGES_REQUESTED** [review]
The logic for decide whether we need to chunk or push the array into the `remset` could be a bit cleaner. I'd refactor:

```c
size_t too_big = (obj_end - obj_begin) / MAX_REFS_AT_ONCE > step; // use this order of operations to avoid idiv
jl_value_t **scan_end = obj_end;
if (too_big) {
    scan_end = obj_begin + step * MAX_REFS_AT_ONCE;
    if ((nptr & 0x3) != 0x2) {
        jl_gc_chunk_t c = {GC_objary_chunk, obj_parent, scan_end,
                           obj_end,      NULL,       NULL,
                           step,         nptr};
        gc_chunkqueue_push(mq, &c);
        obj_end = scan_end;
    }
}
for (; obj_begin < scan_end; obj_begin += step) {
    new_obj = *obj_begin;
    if (new_obj != NULL) {
        verify_parent2("obj array", obj_parent, obj_begin, "elem(%d)",
                       gc_slot_to_arrayidx(obj_parent, obj_begin));
        gc_try_claim_and_push(mq, new_obj, &nptr);
        gc_heap_snapshot_record_array_edge(obj_parent, &new_obj);
    }
}
if (too_big) {
    if (obj_end != scan_end) {
        jl_gc_chunk_t c = {GC_objary_chunk, obj_parent, scan_end,
                           obj_end,      NULL,       NULL,
                           step,         nptr};
        gc_chunkqueue_push(mq, &c);
    }
}
else {
    gc_mark_push_remset(ptls, obj_parent, nptr);
}
```

into:

```c
size_t too_big = (obj_end - obj_begin) / MAX_REFS_AT_ONCE > step; // use this order of operations to avoid idiv
jl_value_t **scan_end = obj_end;
if (too_big) {
    scan_end = obj_begin + step * MAX_REFS_AT_ONCE;
}
for (; obj_begin < scan_end; obj_begin += step) {
    new_obj = *obj_begin;
    if (new_obj != NULL) {
        verify_parent2("obj array", obj_parent, obj_begin, "elem(%d)",
                       gc_slot_to_arrayidx(obj_parent, obj_begin));
        gc_try_claim_and_push(mq, new_obj, &nptr);
        gc_heap_snapshot_record_array_edge(obj_parent, &new_obj);
    }
}
if (too_big) {
    jl_gc_chunk_t c = {GC_objary_chunk, obj_parent, scan_end,
                        obj_end,      NULL,       NULL,
                        step,         nptr};
    gc_chunkqueue_push(mq, &c);
}
else {
    gc_mark_push_remset(ptls, obj_parent, nptr);
}
```

Same change for `gc_mark_array8/16`.

**Comment by vtjnash** [comment]
As long as you don't care that your change eliminates the possibility of intra-array thread-stealing and thus slightly reduces parallelism of threaded GC, yeah, that rearrangement is fine.

**Comment by d-netto** [comment]
> slightly reduces parallelism of threaded GC

This should be fine. If an array is large enough to be chunked then it will expose a lot of `jl_value_t *` that could potentially be stolen.

**Comment by d-netto** [comment]
Also, in https://github.com/JuliaLang/julia/pull/48600 we prioritize stealing pointers before chunks.

File: src/gc.c [hunk 2]
```diff
@@ -2131,7 +2192,17 @@ STATIC_INLINE void gc_mark_array8(jl_ptls_t ptls, jl_value_t *ary8_parent, jl_va
             }
         }
     }
-    gc_mark_push_remset(ptls, ary8_parent, nptr);
+    if (too_big) {
+        if (ary8_end != scan_end) {
+            jl_gc_chunk_t c = {GC_objary_chunk, ary8_parent, scan_end,
+                               ary8_end,      NULL,       NULL,
```
Comment (d-netto) [inline src/gc.c:2198]: Replace `NULL, NULL` here with `elem_begin, elem_end`. Same for `array16`.

File: src/gc.c [hunk 3]
```diff
@@ -2131,7 +2176,15 @@ STATIC_INLINE void gc_mark_array8(jl_ptls_t ptls, jl_value_t *ary8_parent, jl_va
             }
         }
     }
-    gc_mark_push_remset(ptls, ary8_parent, nptr);
+    if (too_big) {
+        jl_gc_chunk_t c = {GC_objary_chunk, ary8_parent, scan_end,
+                            ary8_end,      elem_begin,       elem_end,
+                            elsize,         nptr};
```
Comment (vtjnash) [inline src/gc.c:2182]:
```suggestion
                           ary8_end,        elem_begin,       elem_end,
                           0,               nptr};
```

File: src/gc.c [hunk 4]
```diff
@@ -2143,14 +2196,36 @@ STATIC_INLINE void gc_mark_array16(jl_ptls_t ptls, jl_value_t *ary16_parent, jl_
...
+            if (early_end)
+                break;
+        }
+    }
+    size_t too_big = (ary16_end - ary16_begin) / MAX_REFS_AT_ONCE > elsize; // use this order of operations to avoid idiv
+    jl_value_t **scan_end = ary16_end;
+    if (too_big) {
+        scan_end = ary16_begin + elsize * MAX_REFS_AT_ONCE;
     }
     for (; ary16_begin < ary16_end; ary16_begin += elsize) {
```
Comment (vtjnash) [inline src/gc.c:2230]:
```suggestion
    for (; ary16_begin < scan_end; ary16_begin += elsize) {
```

**Pull request merged**